File: mobile/lib/features/search/search_pair.dart
import 'dart:convert';

import '../../data/local/app_database.dart';

/// Which way the search field is pointing: what the learner types, what they get back.
///
/// A PAIR AND A DIRECTION, carried together, because both halves are asked of the server on every
/// call and because the direction is the thing the pill flips. «EN → RU» and «RU → EN» are the same
/// two languages and two different questions.
class SearchPair {
  const SearchPair({required this.source, required this.target});

  /// What the learner types in.
  final String source;

  /// What comes back.
  final String target;

  /// The same two languages, asked the other way. This is the whole of a tap on the pill.
  SearchPair get swapped => SearchPair(source: target, target: source);

  /// The learner's own side of the pair, given which language the app teaches.
  String otherThan(String taught) => source == taught ? target : source;

  /// True when the query is typed in the learner's own language and the ANSWER is the word being
  /// taught — the direction the small card sets its headline from.
  bool reversedFor(String taught) => target == taught;

  SearchPair withOther(String taught, String other) => reversedFor(taught)
      ? SearchPair(source: other, target: taught)
      : SearchPair(source: taught, target: other);

  Map<String, dynamic> toJson() => {'s': source, 't': target};

  static SearchPair? fromJson(Object? raw) {
    if (raw is! Map) return null;
    final source = (raw['s'] as String?)?.trim() ?? '';
    final target = (raw['t'] as String?)?.trim() ?? '';
    if (source.isEmpty || target.isEmpty || source == target) return null;

    return SearchPair(source: source, target: target);
  }

  @override
  bool operator ==(Object other) =>
      other is SearchPair && other.source == source && other.target == target;

  @override
  int get hashCode => Object.hash(source, target);

  @override
  String toString() => '$source→$target';
}

/// A pair with its ROLES named, which is the shape everything downstream of the search field wants.
///
/// [SearchPair] is a direction — «what I typed» → «what came back» — and it flips twice a minute.
/// A COLLECTION's pair does not flip: it is «изучаемый → язык поддержки» (`target_lang` /
/// `source_lang`), one per collection, forever (DECISIONS п. 81). This is the translation between
/// the two, made once at the top of the save path so no screen further down has to work out which
/// half of the direction was the taught one.
class LearningPair {
  const LearningPair({required this.learned, required this.support});

  /// The language BEING LEARNED — a collection's `target_lang`, and the language of every term in
  /// it. The one half the server's gate actually checks.
  final String learned;

  /// The language of SUPPORT — a collection's `source_lang`: which translation is shown.
  final String support;

  /// Which way round a [SearchPair] is, given the language this deployment teaches.
  factory LearningPair.of(SearchPair pair, String taught) =>
      LearningPair(learned: taught, support: pair.otherThan(taught));

  @override
  bool operator ==(Object other) =>
      other is LearningPair && other.learned == learned && other.support == support;

  @override
  int get hashCode => Object.hash(learned, support);

  @override
  String toString() => '$learned←$support';
}

/// What the server says the pickers may offer, AND which role each language plays.
///
/// ROLES COME FROM HERE AND NOWHERE ELSE. `lib/l10n/language_endonyms.dart` is a table of NAMES —
/// it says what a language is called and what flag it flies, not whether this deployment can teach
/// it or read it. Those two lists are a capability of the server (`GET /search/languages`), they
/// change by deployment rather than by release, and a client that held its own copy would offer a
/// pair the server then refuses. So the «На какой» picker lists [learned] and the «С какого» picker
/// lists [support], and neither is written out in the app.
///
/// [learned] HOLDS EXACTLY ONE ENTRY TODAY, because the server names exactly one: `SupportedLanguages`
/// still requires the taught language on one side of every pair (`config/languages.php`), which
/// DECISIONS п. 134 records as a v1 product limit due to be lifted (RS-3), not as the rule. The
/// picker is written against a list so that lifting it on the server is the whole change.
class SearchLanguages {
  const SearchLanguages({required this.taught, required this.natives, required this.defaultNative});

  /// The language the app teaches — the fixed half of every pair in v1.
  final String taught;

  /// The languages a learner may read, in the order to offer them.
  final List<String> natives;

  /// Where the pill starts on a device that has never been set: the learner's own language.
  final String defaultNative;

  /// What may occupy the LEARNED side of the pair — «На какой».
  List<String> get learned => [taught];

  /// What may occupy the SUPPORT side of the pair — «С какого».
  List<String> get support => natives;

  /// Which language may stand in the slot [code] currently occupies. The two pills are the two
  /// SIDES of the direction, so the swap moves the roles between them and each pill has to ask
  /// again which list it is offering.
  List<String> optionsAt(String code) => code == taught ? learned : support;

  /// The pill's opening position — taught language into their own. Looking up a word they have
  /// just met is the commoner half of the job, so that is the way round it opens.
  SearchPair get initialPair => SearchPair(source: taught, target: defaultNative);

  factory SearchLanguages.fromJson(Map<String, dynamic> j) {
    final taught = (j['target'] as String?)?.trim();
    final natives = [for (final n in (j['natives'] as List? ?? const [])) (n as String).trim()]
      ..removeWhere((n) => n.isEmpty);

    return SearchLanguages(
      taught: (taught ?? '').isNotEmpty ? taught! : 'en',
      natives: natives.isNotEmpty ? natives : const ['ru'],
      defaultNative: (j['default_native'] as String?)?.trim().isNotEmpty == true
          ? (j['default_native'] as String).trim()
          : (natives.isNotEmpty ? natives.first : 'ru'),
    );
  }
}

/// The chosen pair, kept on the device and nowhere else.
///
/// LOCAL because it is a preference about a screen, not progress: which way somebody last pointed
/// their search field is not worth a column on the server, and it should survive a restart of the
/// app rather than a re-login. It rides the same sync-meta key/value table as the search history,
/// so it needs no migration and it leaves with the account when the database is cleared.
///
/// Every read is defensive. A key written by an older build, hand-edited, or naming a language the
/// server has since stopped serving must degrade to «the default pair», never to a search the
/// server will refuse — which is exactly what a remembered `ro` would become if `ro` were dropped
/// from the deployment.
class SearchPairStore {
  const SearchPairStore(this._db);

  final AppDatabase _db;

  static const metaKey = 'search.pair';

  /// The stored pair if it is still one the server serves, otherwise [languages]' opening position.
  Future<SearchPair> load(SearchLanguages languages) async {
    final raw = await _db.getMeta(metaKey);
    if (raw == null || raw.isEmpty) return languages.initialPair;
    try {
      final stored = SearchPair.fromJson(jsonDecode(raw));
      if (stored == null || !_serves(languages, stored)) return languages.initialPair;

      return stored;
    } catch (_) {
      return languages.initialPair;
    }
  }

  Future<void> save(SearchPair pair) => _db.setMeta(metaKey, jsonEncode(pair.toJson()));

  /// Exactly one side taught, the other a language still on offer — the same rule the server
  /// enforces, applied here so a stale preference never becomes a 422 the learner has to read.
  static bool _serves(SearchLanguages languages, SearchPair pair) {
    final taught = languages.taught;
    final sides = {pair.source, pair.target};

    return sides.contains(taught) &&
        sides.length == 2 &&
        languages.natives.contains(pair.otherThan(taught));
  }
}

File: backend2/app/Modules/Generation/Application/Service/SearchPair.php
<?php

declare(strict_types=1);

namespace App\Modules\Generation\Application\Service;

use App\Modules\Generation\Application\Dto\ResolvedPair;
use App\Modules\Generation\Domain\Exception\TaughtSideNotTaught;
use App\Modules\Generation\Domain\Exception\UnsupportedLanguagePair;
use App\Modules\Generation\Domain\Service\SearchDirection;
use App\Modules\Generation\Domain\Service\SupportedLanguages;
use App\Modules\Generation\Domain\ValueObject\TaughtSide;
use App\Modules\Shared\Domain\Service\LanguageRoles;
use App\Modules\Shared\Domain\ValueObject\UserId;

/**
 * The pair a search runs in, resolved once for all three endpoints.
 *
 * ## The learner says which way, and nothing else gets a vote
 *
 * This replaces an earlier design where the vendor's own language detection decided the direction.
 * It was overruled for a concrete reason: on single words the detector is unreliable in a way that
 * is invisible until it bites — «gate» comes back as Norwegian and is answered as «улица», «случай»
 * as Bulgarian. A detector that is right nine times in ten is worse than no detector on a screen
 * where the tenth answer is a word the learner then studies.
 *
 * So the direction comes from the pill on the search screen, the pill remembers what it was last
 * set to, and the server takes it as given. The provider is called with an explicit source and an
 * explicit target every time.
 *
 * ## Falling back
 *
 * A request that names no pair gets the learner's profile one — English and their own language.
 * That is not a nicety for old clients only: `GET /search` is also reached from places that have no
 * pill, and «what this learner reads» is the right default everywhere.
 *
 * A pair that IS named and is not supported is refused ({@see UnsupportedLanguagePair}), never
 * quietly swapped for the default — see that class for why.
 *
 * ## Which side is being learned
 *
 * A direction is not a pair of roles, and the client may now say the roles outright
 * ({@see TaughtSide}). When it does not, {@see taughtSideOf()} works them out — from the pair when
 * only one side is teachable, and from the direction when both are. The profile does not get a
 * vote: it decides the OPENING pair and nothing else (DECISIONS пп. 142, 147).
 */
final readonly class SearchPair
{
    public function __construct(
        private LearnerLanguages $languages,
        private SupportedLanguages $supported,
    ) {}

    /**
     * @param  string|null  $source  the language the query is written in
     * @param  string|null  $target  the language the answer comes back in
     * @param  TaughtSide|null  $taughtSide  which half of the direction the learner is STUDYING,
     *                                       said by the client. Null falls back to `taughtSideOf()`.
     *
     * @throws UnsupportedLanguagePair
     * @throws TaughtSideNotTaught
     */
    public function resolve(
        UserId $actorId,
        ?string $source,
        ?string $target,
        ?TaughtSide $taughtSide = null,
    ): ResolvedPair {
        $from = strtolower(trim($source ?? ''));
        $to = strtolower(trim($target ?? ''));

        // Half a pair is not a pair. Both or neither — a request that named only a target would
        // otherwise get a direction nobody chose.
        if ($from === '' || $to === '') {
            return $this->fromProfile($actorId);
        }

        if (! $this->supported->supports($from, $to)) {
            throw UnsupportedLanguagePair::of($from, $to);
        }

        // A supported pair has a taught language on one side — but «which side» is a question the
        // direction alone answers only when the OTHER side is not taught too. The client may now
        // ANSWER it instead of leaving it to be guessed; `taughtSideOf()` is the fallback for the
        // requests that do not (DECISIONS п. 147).
        $termLang = $taughtSide !== null
            ? $this->namedTaughtSide($taughtSide, $from, $to)
            : $this->taughtSideOf($from, $to);

        return new ResolvedPair(
            direction: new SearchDirection($from, $to),
            termLang: $termLang,
            translationLang: $termLang === $from ? $to : $from,
        );
    }

    /**
     * The side the CLIENT named, checked against what this deployment can actually teach.
     *
     * Checked rather than trusted: a role is a claim about the deployment's capabilities, and a
     * client that has not re-read `GET /search/languages` since a language was retired would
     * otherwise have the server answer in a language nobody teaches — invisibly, because the screen
     * would look exactly the same.
     *
     * @throws TaughtSideNotTaught
     */
    private function namedTaughtSide(TaughtSide $side, string $from, string $to): string
    {
        $lang = $side->of($from, $to);
        if (! $this->supported->teaches($lang)) {
            throw TaughtSideNotTaught::of($side, $lang);
        }

        return $lang;
    }

    /**
     * Which half of the direction is the language being LEARNED.
     *
     * Usually the pair names it: in `ru → ro` only one side is a language this product teaches, and
     * that is the term side whichever way the learner typed it — `ru` has no trainers, so nobody is
     * learning it here. That branch is {@see SupportedLanguages::soleTaughtSide()} and it is most
     * requests.
     *
     * `en → es` names TWO taught languages, and there the DIRECTION is the answer: the learner types
     * what they already have and asks for what they do not. «Translate this English word into
     * Spanish» is somebody studying Spanish, so the TARGET side is the taught one and the source is
     * support. The same reading run backwards is just as true: `es → en` is somebody studying
     * English who happens to write Spanish.
     *
     * THIS REPLACED A PROFILE TIE-BREAK (DECISIONS п. 147, amended 24.08), and the replacement is a
     * bug fix rather than a preference. The profile rule answered `en → es` with «English studied
     * with Spanish support», because English is what `profiles.default_target_lang` holds — so the
     * card put the English word in the headline, the Spanish in small type underneath, and the save
     * sheet offered to file the word under «English ← Spanish». Every one of those is the opposite
     * of what the learner asked for, and none of them looks broken on screen: it looks like an app
     * that disagrees with you. A rule that only holds while somebody studies exactly one language
     * was never going to survive a product whose whole current chapter is «any pair» (п. 134).
     *
     * The profile keeps the one job it is allowed to keep — the OPENING pair of the screen
     * ({@see fromProfile()}, п. 142). It no longer decides what a stated pair means.
     */
    private function taughtSideOf(string $from, string $to): string
    {
        $sole = $this->supported->soleTaughtSide($from, $to);

        return $sole ?? LanguageRoles::normalize($to);
    }

    /**
     * The learner's own pair, as their profile has it.
     *
     * The sides come from the PROFILE, which is where a request that named no pair has to get them:
     * this is the «стартовая пара поиска» the profile is still allowed to decide (DECISIONS п. 142).
     */
    public function fromProfile(UserId $actorId): ResolvedPair
    {
        $langs = $this->languages->forUser($actorId);

        return new ResolvedPair(
            // Taught language first: «EN → RU» is the direction somebody opens the screen in,
            // because looking up a word they have just met is the commoner half of the job.
            direction: new SearchDirection($langs->target->value, $langs->native->value),
            termLang: $langs->target->value,
            translationLang: $langs->native->value,
        );
    }
}
